### PR TITLE
payment failure message with attached channel_update not validated

### PR DESCRIPTION
**Pull request by joostjager**
The intention of this PR is to fix https://github.com/lightningnetwork/lnd/issues/1707.

### CODE REVIEW

File: routing/router_test.go [hunk 1]
```diff
@@ -347,6 +347,163 @@ func TestSendPaymentRouteFailureFallback(t *testing.T) {
...
+
+	testGraph, err := createTestGraphFromChannels(testChannels)
+	defer testGraph.cleanUp()
+	if err != nil {
+		t.Fatalf("unable to create graph: %v", err)
+	}
+
+	const startingBlockHeight = 101
+
+	ctx, cleanUp, err := createTestCtxFromGraphInstance(startingBlockHeight, testGraph)
```
Comment (halseth) [inline routing/router_test.go]: nit: long line

File: routing/router_test.go [hunk 2]
```diff
@@ -347,6 +347,163 @@ func TestSendPaymentRouteFailureFallback(t *testing.T) {
...
+	}
+
+	// The payment parameter is mostly redundant in SendToRoute. Can be left
+	// empty for this test.
+	payment := &LightningPayment{}
+
+	// Send off the payment request to the router. The specified route
+	// should be attempted and the channel update should be received by
+	// router and ignored because it is missing a valid signature.
+	_, _, _ = ctx.router.SendToRoute([]*Route{route}, payment)
```
Comment (halseth) [inline routing/router_test.go]: check error

File: routing/router_test.go [hunk 3]
```diff
@@ -347,6 +347,163 @@ func TestSendPaymentRouteFailureFallback(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	errChanUpdate.Signature, err = lnwire.NewSigFromSignature(sig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Retry the payment using the same route as before.
+	_, _, _ = ctx.router.SendToRoute([]*Route{route}, payment)
```
Comment (halseth) [inline routing/router_test.go]: check error

File: routing/router_test.go [hunk 4]
```diff
@@ -14,6 +14,7 @@ import (
 	"github.com/lightningnetwork/lnd/htlcswitch"
 
 	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
```
Comment (halseth) [inline routing/router_test.go]: belongs in a different commit?

File: discovery/gossiper.go [hunk 5]
```diff
@@ -1519,7 +1519,7 @@ func (d *AuthenticatedGossiper) processRejectedEdge(chanAnnMsg *lnwire.ChannelAn
 	if err != nil {
 		return nil, err
 	}
-	err = ValidateChannelAnn(chanAnn)
+	err = routing.ValidateChannelAnn(chanAnn)
```
Comment (halseth) [inline discovery/gossiper.go]: add an explanation to the commit message the rationale for moving it.

File: routing/router.go [hunk 6]
```diff
@@ -2003,13 +2009,18 @@ func pruneEdgeFailure(paySession *paymentSession, route *Route,
 
 // applyChannelUpdate applies a channel update directly to the database,
 // skipping preliminary validation.
-func (r *ChannelRouter) applyChannelUpdate(msg *lnwire.ChannelUpdate) error {
+func (r *ChannelRouter) applyChannelUpdate(msg *lnwire.ChannelUpdate,
```
Comment (halseth) [inline routing/router.go]: godoc mus be updated.

File: routing/router_test.go (hunk 4 above)
Comment (joostjager) [reply routing/router_test.go]: indeed

File: routing/pathfind_test.go [hunk 7]
```diff
@@ -329,26 +328,36 @@ type testChannel struct {
 	Capacity btcutil.Amount
 }
 
-// createTestGraph returns a fully populated ChannelGraph based on a set of
+type testGraphInstance struct {
```
Comment (halseth) [inline routing/pathfind_test.go]: 👍

File: routing/pathfind_test.go [hunk 8]
```diff
@@ -466,13 +475,16 @@ func createTestGraph(testChannels []*testChannel) (*channeldb.ChannelGraph, func
...
 		}
 
 		channelID++
 	}
 
-	return graph, cleanUp, aliasMap, nil
+	return &testGraphInstance{
+		graph:    graph,
+		cleanUp:  cleanUp,
+		aliasMap: aliasMap}, nil
```
Comment (halseth) [inline routing/pathfind_test.go]: style nit: newline before closing bracket

File: routing/router_test.go [hunk 9]
```diff
@@ -347,6 +348,171 @@ func TestSendPaymentRouteFailureFallback(t *testing.T) {
...
+
+	// The payment parameter is mostly redundant in SendToRoute. Can be left
+	// empty for this test.
+	payment := &LightningPayment{}
+
+	// Send off the payment request to the router. The specified route
+	// should be attempted and the channel update should be received by
+	// router and ignored because it is missing a valid signature.
+	_, _, err = ctx.router.SendToRoute([]*Route{route}, payment)
+	if err == nil {
```
Comment (halseth) [inline routing/router_test.go]: can you check that the expected error is returned?

File: routing/router_test.go [hunk 10]
```diff
@@ -347,6 +348,171 @@ func TestSendPaymentRouteFailureFallback(t *testing.T) {
...
+	}
+
+	errChanUpdate.Signature, err = lnwire.NewSigFromSignature(sig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Retry the payment using the same route as before.
+	_, _, err = ctx.router.SendToRoute([]*Route{route}, payment)
+	if err == nil {
```
Comment (halseth) [inline routing/router_test.go]: same

**Comment by halseth** [comment]
Also needs a rebase.

File: routing/router_test.go (hunk 9 above)
Comment (joostjager) [reply routing/router_test.go]: do you want me to check on strings?

**Comment by joostjager** [comment]
Rebased
